tools/halide_malloc_trace.h

```diff
@@ -0,0 +1,102 @@
+#ifndef HALIDE_MALLOC_TRACE_H
+#define HALIDE_MALLOC_TRACE_H
+
+//---------------------------------------------------------------------------
+// The custom trace allocator can be used in an application by calling:
+//
+//   halide_enable_malloc_trace();
+//
+// When the app is run, calls to halide_malloc/free will produce output like:
+//
+//   halide_malloc => [0x9e400, 0xa27ff], # size:17408, align:1K
+//   halide-header => [0x9e390, 0x9e3ff], # size:112, align:16
+//   halide_malloc => [0xa2880, 0xa6e9f], # size:17952, align:128
+//   halide-header => [0xa2820, 0xa287f], # size:96, align:32
+//   halide_free   => [0x9e390, 0x9e3ff], # size:112, align:16
+//   halide_free   => [0xa2820, 0xa287f], # size:96, align:32
+//
+//---------------------------------------------------------------------------
+
+#include <cstdlib>
+#include <memory>
+#include <iostream>
+
+namespace Halide {
+namespace Tools {
+
+static inline void print_meminfoalign(intptr_t val) {
+    intptr_t align_chk = 1024*1024;
+    while (align_chk > 0) {
+        if ((val & (align_chk-1)) == 0) {
+            char aunit = ' ';
+            if (align_chk >= 1024) {
+                align_chk >>= 10;
+                aunit = 'K';
+            }
+            if (align_chk >= 1024) {
+                align_chk >>= 10;
+                aunit = 'M';
+            }
+            std::cout << "align:" << align_chk;
+            if (aunit != ' ') {
+                std::cout << aunit;
+            }
+            break;
+        }
+        align_chk >>= 1;
+    }
+}
+
+void *halide_malloc_trace(void *user_context, size_t x) {
+    // Halide requires halide_malloc to allocate memory that can be
+    // read 8 bytes before the start and 8 bytes beyond the end.
+    // Additionally, we also need to align it to the natural vector
+    // width.
+    void *orig = malloc(x+(128+8));
```

Comment:
#1046 will make this +8 unnecessary, since the internal codegen takes care of it.